extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergCatalogAdapter.java

```diff
@@ -826,8 +798,500 @@ private Table readTableInternal(
+            }
+            fieldID++;
+        }
+        final Schema schema = new Schema(fields);
+
+        final PartitionSpec partitionSpec = createPartitionSpec(schema, partitioningColumnNames);
+        return new SpecAndSchema(partitionSpec, schema);
+    }
+
+    private static PartitionSpec createPartitionSpec(
+            @NotNull final Schema schema,
+            @NotNull final Iterable<String> partitionColumnNames) {
+        final PartitionSpec.Builder partitionSpecBuilder = PartitionSpec.builderFor(schema);
+        for (final String partitioningColumnName : partitionColumnNames) {
+            partitionSpecBuilder.identity(partitioningColumnName);
+        }
+        return partitionSpecBuilder.build();
+    }
+
+    private org.apache.iceberg.Table createNewIcebergTable(
+            @NotNull final TableIdentifier tableIdentifier,
+            @NotNull final SpecAndSchema specAndSchema,
+            @NotNull final IcebergParquetWriteInstructions writeInstructions) {
+        if (catalog.tableExists(tableIdentifier)) {
+            throw new IllegalArgumentException("Table already exists: " + tableIdentifier);
+        }
+        return catalog.createTable(tableIdentifier, specAndSchema.schema, specAndSchema.partitionSpec, null,
+                Map.of(TableProperties.DEFAULT_FILE_FORMAT, DEFAULT_GENERATED_FILE_FORMAT,
+                        TableProperties.PARQUET_COMPRESSION,
+                        StringUtils.toLowerCase(writeInstructions.compressionCodecName())));
+    }
+
+    private static class CompletedParquetWrite {
+        private final URI destination;
+        private final long numRows;
+        private final long numBytes;
+
+        private CompletedParquetWrite(final URI destination, final long numRows, final long numBytes) {
+            this.destination = destination;
+            this.numRows = numRows;
+            this.numBytes = numBytes;
+        }
+    }
+
+    @NotNull
+    private static List<CompletedParquetWrite> writeParquet(
+            @NotNull final org.apache.iceberg.Table icebergTable,
+            @NotNull final Table[] dhTables,
+            @NotNull final IcebergParquetWriteInstructions writeInstructions) {
+        // Build the parquet instructions
+        final List<CompletedParquetWrite> parquetFilesWritten = new ArrayList<>(dhTables.length);
+        final ParquetInstructions.OnWriteCompleted onWriteCompleted =
+                (destination, numRows, numBytes) -> parquetFilesWritten
+                        .add(new CompletedParquetWrite(destination, numRows, numBytes));
+        final ParquetInstructions parquetInstructions = writeInstructions.toParquetInstructions(
+                onWriteCompleted, icebergTable.schema().idToName());
+
+        // Write the data to parquet files
+        for (final Table dhTable : dhTables) {
+            final long epochMicrosNow = Instant.now().getEpochSecond() * 1_000_000 + Instant.now().getNano() / 1_000;
```

Comment:
This is not actually _guaranteed_ to be increasing; it's possible that time can go backwards. If we want to ensure these files get written with increasing names we should:
1. Take a single timestamp outside the for loop
2. Add a (properly padded) index

But even that is not enough if there are going to be other writers after this.